src/net_doh.c

```diff
@@ -0,0 +1,1140 @@
+    }
+
+    if (self->is_ready &&
+        self->http2 != NULL) {
+        // do nghttp2 I/O send to flush outstanding frames
+        int ret;
+    
+        ret = nghttp2_session_send(self->http2->session);
+        if (ret != 0) {
+            perf_log_printf("nghttp2_session_send failed: %s", nghttp2_strerror(ret));
+            http2_session_free(sock);
+            PERF_UNLOCK(&self->lock);
+            return 0;
+        }
+
+        PERF_UNLOCK(&self->lock);
+        return 1;
+    }
+
+    if (!self->is_conn_ready) {
+        switch (perf_os_waituntilanywritable(&sock, 1, pipe_fd, timeout)) {
+        case PERF_R_TIMEDOUT:
+            PERF_UNLOCK(&self->lock);
+            return -1;
+        case PERF_R_SUCCESS: {
+            int       error = 0;
+            socklen_t len   = (socklen_t)sizeof(error);
+
+            getsockopt(sock->fd, SOL_SOCKET, SO_ERROR, (void*)&error, &len);
+            if (error != 0) {
+                if (error == EINPROGRESS
+#if EWOULDBLOCK != EAGAIN
+                    || error == EWOULDBLOCK
+#endif
+                    || error == EAGAIN) {
+                    PERF_UNLOCK(&self->lock);
+                    return 0;
+                }
+                // unrecoverable error, reconnect
+                self->do_reconnect = true;
+                PERF_UNLOCK(&self->lock);
+                return 0;
+            }
+            break;
+        }
+        default:
+            PERF_UNLOCK(&self->lock);
+            return -1;
+        }
+        self->is_conn_ready = true;
+    }
+
+    int ret = SSL_connect(self->ssl);
+    if (!ret) {
+        // unrecoverable error, reconnect
+        self->do_reconnect = true;
+        PERF_UNLOCK(&self->lock);
+        return 0;
+    }
+    if (ret < 0) {
+        switch (SSL_get_error(self->ssl, ret)) {
+        case SSL_ERROR_WANT_READ:
+        case SSL_ERROR_WANT_WRITE:
+            break;
+        default:
+            // unrecoverable error, reconnect
+            self->do_reconnect = true;
+        }
+        PERF_UNLOCK(&self->lock);
+        return 0;
+    }
+
+    self->is_ready = true;
+
+    const uint8_t *alpn = NULL;
+    uint32_t alpn_len = 0;
+ #ifndef OPENSSL_NO_NEXTPROTONEG
+    SSL_get0_next_proto_negotiated(self->ssl, &alpn, &alpn_len);
+#endif /* !OPENSSL_NO_NEXTPROTONEG */
+#if OPENSSL_VERSION_NUMBER >= 0x10002000L
+    if (alpn == NULL) {
+        SSL_get0_alpn_selected(self->ssl, &alpn, &alpn_len);
+    }
+#endif /* OPENSSL_VERSION_NUMBER >= 0x10002000L */
+
+    if (alpn == NULL || 
+        alpn_len != 2 || 
+        memcmp("h2", alpn, 2) != 0) {
+        http2_session_free(sock);
+        perf_log_fatal("failed to negotiate h2 proto.");
+    }
+
+    // guard against re-entrant http2_init
+    if (!self->http2->settings_sent) {
+        // send settings
+        ret = _http2_send_settings(self->http2);
+        if (ret != 0) {
+            perf_log_printf("nghttp2_submit_settings failed: %s", nghttp2_strerror(ret));
```

Comment:
this must be handled better, can settings be resent? does this warrant a reconnect?

Reply:
@jelu: you are right, it does. reconnect might succeed re-establishing new http2 session.

I had a wrong initial assumption on my side that new query opens new socket, does the job and closes it and haven't looked explicitly on recovery mechanisms, but it clear now after your explanation and I will add reconnects where due.

Reply:
and on http2 settings resent - we can to a retry of nghttp2_submit_settings() on failure, but no guarantees that subsequent attempt will not fail i.e. cannot try forever. I will add reconnect here.